crates/database/src/leader_pg_lock.rs

```diff
@@ -0,0 +1,89 @@
+use {
+    sqlx::postgres::PgPool,
+    std::time::{Duration, Instant},
+};
+
+struct PgLockGuard {
+    conn: sqlx::pool::PoolConnection<sqlx::Postgres>,
+    key: String,
+}
+
+impl PgLockGuard {
+    async fn ping(&mut self) -> bool {
+        sqlx::query("SELECT 1")
+            .execute(&mut *self.conn)
+            .await
+            .is_ok()
+    }
+
+    async fn unlock(mut self) {
+        // TODO what can we do with the result?
+        let _ = sqlx::query("SELECT pg_advisory_unlock($1)")
+            .bind(self.key)
+            .execute(&mut *self.conn)
+            .await
+            .map_err(|err| {
+                tracing::warn!(error = %err, "lock release failed");
+            });
+        // conn returns to pool unlocked
+    }
+}
+
+pub struct LeaderLock {
+    pool: PgPool,
+    key: String,
+    lock_guard: Option<PgLockGuard>,
+    last_try: Instant,
+    try_every: Duration,
+}
+
+impl LeaderLock {
+    pub fn new(pool: PgPool, key: String, try_every: Duration) -> Self {
+        Self {
+            pool,
+            key,
```

Comment:
Is there a reason to pass the key as an argument? We wouldn't want different locks to be instantiated by accident so hardcoding a value here seems less error prone.

Reply:
The idea is that there might be other critical sections one wants to use this for, e.g. in other services that need synchronization of this kind. So far we have one use-case so this is strictly speaking not needed, yes.

Reply:
I do support @fafk rationale here. On top of that, the key being a custom value is natively exposed by Postgres. If we were to hardcode it then maybe a safer approach would be to have generic const, but that would increase the churn related to propagating the presence of generic constant to the structs that use it.